hnn_core/dipole.py

```diff
@@ -152,64 +194,133 @@ class Dipole(object):
         for key in self.data.keys():
             self.data[key] *= fctr
-        return fctr
+        return self
+
+    def smooth(self, window_len):
+        """Smooth the dipole waveform using Hamming-windowed convolution
+
+        Note that this method operates in-place, i.e., it will alter the data.
+        If you prefer a filtered copy, consider using the
+        `~hnn_core.dipole.copy`-method.
+
+        Parameters
+        ----------
+        window_len : float or None
+            The length (in ms) of a `~numpy.hamming` window to convolve the
+            data with.
+
+        Returns
+        -------
+        dpl_copy : instance of Dipole
+            A copy of the modified Dipole instance.
+        """
+        # convolutional filter length is given in samples
+        winsz = np.round(1e-3 * window_len * self.sfreq)
+        if winsz > len(self.times):
+            raise ValueError(
+                f'Window length too long: {winsz} samples; data length is '
+                f'{len(self.times)} samples')
+        elif winsz > 0 and winsz < 1:
+            raise ValueError('Window length less than 1 ms is not supported')
+        elif winsz < 0:
+            raise ValueError('Window length cannot be negative')
 
-    def smooth(self, winsz):
-        # XXX: add check to make sure self.times is
-        # not smaller than winsz
-        if winsz <= 1:
-            return
         for key in self.data.keys():
             self.data[key] = _hammfilt(self.data[key], winsz)
 
+        return self
+
+    def savgol_filter(self, h_freq):
+        """Smooth the dipole waveform using Savitzky-Golay filtering
+
+        Note that this method operates in-place, i.e., it will alter the data.
+        If you prefer a filtered copy, consider using the
+        `~hnn_core.dipole.copy`-method. The high-frequency cutoff value of a
+        Savitzky-Golay filter is approximate; see `~scipy.signal.savgol_filter.
+
+        Parameters
+        ----------
+        h_freq : float or None
+            Approximate high cutoff frequency in Hz. Note that this
+            is not an exact cutoff, since Savitzky-Golay filtering
+            is done using polynomial fits
+            instead of FIR/IIR filtering. This parameter is thus used to
+            determine the length of the window over which a 5th-order
+            polynomial smoothing is applied.
+
+        Returns
+        -------
+        dpl_copy : instance of Dipole
+            A copy of the modified Dipole instance.
+        """
+        if h_freq < 0:
+            raise ValueError('h_freq cannot be negative')
+        elif h_freq > 0.5 * self.sfreq:
+            raise ValueError(
+                'h_freq must be less than half the sample rate')
+        for key in self.data.keys():
+            self.data[key] = _savgol_filter(self.data[key],
+                                            h_freq,
+                                            self.sfreq)
+        return self
```

Comment:
Should this be an optional return? I get the usefulness for chaining functions, but stating that the function acts in place would lead me to expect that it has no return value.

Reply:
Hmm, dunno. @jasmainak  do software design best-principles help to resolve this?